traffic_ops/traffic_ops_golang/deliveryservice/consistenthash/consistenthash.go

```diff
@@ -0,0 +1,131 @@
+package consistenthash
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+)
+
+// struct for the response object from Traffic Router
+type TRConsistentHashResult struct {
+	ResultingPathToConsistentHash string `json:"resultingPathToConsistentHash"`
+	ConsistentHashRegex           string `json:"consistentHashRegex"`
+	RequestPath                   string `json:"requestPath"`
+}
+
+// struct for the incoming request object
+type TRConsistentHashRequest struct {
+	ConsistentHashRegex string `json:"regex"`
+	RequestPath         string `json:"requestPath"`
+	CdnID               int64  `json:"cdnId"`
+}
+
+// endpoint to test Traffic Router's Pattern-Based Consistent Hashing feature
+// /api/1.5/consistenthash
+func Post(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	req := TRConsistentHashRequest{}
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("malformed JSON: "+err.Error()), nil)
+		return
+	}
+
+	responseFromTR, err := getPatternBasedConsistentHash(inf.Tx.Tx, req.ConsistentHashRegex, req.RequestPath, req.CdnID)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("getting pattern based consistent hash from Traffic Router: "+err.Error()))
+		return
+	}
+
+	consistentHashResult := TRConsistentHashResult{}
+	json.Unmarshal(responseFromTR, &consistentHashResult)
+
+	api.WriteResp(w, r, consistentHashResult)
+}
+
+const RouterRequestTimeout = time.Second * 10
+
+// queries database for active Traffic Router on the CDN specified by cdnId
+// passes regex and requestPath to the Traffic Router via API request,
+// and returns the response
+func getPatternBasedConsistentHash(tx *sql.Tx, regex string, requestPath string, cdnId int64) ([]byte, error) {
+	q := `
```

Comment:
I'm not sure based on the examples set forth by extant code, but shouldn't this be a `const`? Seems like allocating space for this on a per-call basis might be a waste of threaded stack memory and CPU cycles (though maybe the compiler is smart enough to make that optimization anyway?).